source/agora/consensus/protocol/Nominator.d

```diff
@@ -526,8 +665,102 @@ extern(D):
+        // if we're ready to confirm then derive the block and sign its hash
+        if (envelope.statement.pledges.type_ == SCPStatementType.SCP_ST_CONFIRM)
+            this.signConfirmBallot(envelope);
+
+        const Scalar challenge = SCPStatementHash(&envelope.statement).hashFull();
+        envelope.signature = sign(this.schnorr_pair, challenge);
+        log.trace("SIGN Envelope signature {} \nfor public key {} \n" ~
+                "envelope {}\n challenge {}",
+                envelope.signature, public_key,
+                scpPrettify(&envelope), challenge);
+    }
+
+    /***************************************************************************
+
+        Signs the confirm ballot in the SCPEnvelope
+
+        Params:
+            envelope = the SCPEnvelope
+
+    ***************************************************************************/
+
+    private void signConfirmBallot (ref SCPEnvelope envelope) nothrow
+    {
+        assert(envelope.statement.pledges.type_ ==
+            SCPStatementType.SCP_ST_CONFIRM);
+
+        ConsensusData con_data;
+        try
+        {
+            con_data = deserializeFull!ConsensusData(
+                envelope.statement.pledges.confirm_.ballot.value[]);
+        }
+        catch (Exception ex)
+        {
+            assert(0);  // this should never happen
+        }
+
+        auto blocks = this.ledger.getBlocksFrom(
+            Height(envelope.statement.slotIndex - 1));
+        assert(!blocks.empty);
+
+        const prev_block = blocks.front;
+        const proposed_block = makeNewBlock(prev_block, con_data.tx_set,
+            con_data.enrolls);
+
+        Height height = proposed_block.header.height;
+        const Signature signature = createBlockSignature(proposed_block);
+
+        envelope.statement.pledges.confirm_.value_sig = opaque_array!64(BitBlob!512(signature[]));
+        auto block_sig = ValidatorBlockSig(height, public_key, signature);
+
+        // Store our block signature in the slot_sigs map
+        log.trace("signConfirmBallot: ADD block signature at height {} for this node {}",
+            height, this.public_key);
+        this.slot_sigs[block_sig.height][this.schnorr_pair.V] = signature;
+    }
+
+    /***************************************************************************
+
+        Add the block signature for a confirmed ballot.
+        The passed envelope's signature must already be verified,
+        and the envelope statement must be a confirmed ballot.
+
+        Params:
+            public_key = the key which signed this envelope,
+                needed for proposed block header signature validation
+            block = the proposed block
+            signature = block signature to collect
+
+        Returns:
+            true if the SCP protocol accepted this envelope
+
+    ***************************************************************************/
+
+    private bool addBlockSignature (const ref PublicKey public_key, const Block block,
```

Comment:
Now we have a parameter named `public_key` as well as `this.public_key`. If these are not the same they should be named differently. If they are the same then we don't need the parameter.

Reply:
I'll rename the node's public_key to node_public_key